### PR TITLE
perf: float32 viability experiment for JaxEngine (#178)

### LINKED ISSUE #178: perf: Float32 viability experiment for JaxEngine harmonic force fields
## Summary

Test whether float32 is viable for JaxEngine's harmonic-only force fields. If it is, consumer GPUs like the RTX 5090 unlock 64× more throughput (104.8 vs 1.6 TFLOPS).

## Background

q2mm currently forces `jax_enable_x64=True` globally in `_jax_common.py`. [GPU benchmarks](https://ericchansen.github.io/q2mm/benchmarks/gpu/) show CPU beating GPU by 1.6–5.5×, and the float64 penalty on consumer GPUs is the dominant factor.

Numerical analysis suggests float64 may not be necessary for the current harmonic-only JaxEngine:

- The unit conversion (kcal/mol/Å² → Hartree/Bohr²) is **linear** — it does not change relative precision
- For harmonic terms (`E = k(r − r₀)²`), the Hessian is exactly `2k` — no catastrophic cancellation in the autodiff tape
- To resolve 1 cm⁻¹ at a 1000 cm⁻¹ mode requires ~8% relative precision; float32 provides 1.2 × 10⁻⁷ — **5 orders of magnitude** more than needed

Float64 would be needed for:
- VdW terms (1/r¹² − 1/r⁶ near-cancellation)
- Morse potentials, cross-terms
- Very soft modes (~10 cm⁻¹)

## Experiment

1. Run CH₃F and rh-enamide benchmarks with `jax_enable_x64=False`
2. Compare computed frequencies against float64 baseline
3. Measure GPU throughput improvement
4. Document results in `docs/benchmarks/gpu.md`

## Success criteria

- [ ] Frequency comparison table (float32 vs float64) for both systems
- [ ] GPU benchmark numbers with float32
- [ ] Clear recommendation: viable / not viable / conditionally viable
- [ ] If viable, document which force field terms are safe in float32

## Related

- #176 (end-to-end differentiable loss — this experiment informs Phase 0)
- #155 (GPU acceleration umbrella)

### PR DESCRIPTION
**Pull request by ericchansen**
### Summary

Investigated whether float32 precision is viable for JaxEngine's harmonic-only force fields, which would unlock 64× throughput on consumer GPUs (FP32:FP64 ratio).

### Result: **Not viable** for molecules >30 atoms

| System | Atoms | Max Δ (real modes) | Threshold | Verdict |
|--------|-------|--------------------|-----------|---------|
| CH3F | 5 | 0.0002 cm⁻¹ | 0.1 cm⁻¹ | ✅ Pass |
| Rh-enamide | 62 | 0.78 cm⁻¹ | 0.1 cm⁻¹ | ❌ Fail |

**Bottleneck:** Eigenvalue decomposition (`jnp.linalg.eigh`) of 186×186 Hessians loses precision in float32. The energy computation itself is fine — it's the frequency extraction step that breaks.

### Changes

1. **`q2mm/backends/mm/_jax_common.py`** — Respect `JAX_ENABLE_X64=0` env var instead of always overriding to True. Enables reproducible float32 experiments.
2. **`scripts/float32_experiment.py`** — Reproducible experiment script with `run64`/`run32`/`compare` subcommands.
3. **`docs/benchmarks/gpu.md`** — Updated float32 section with experiment results, float64 discussion, and JAX-ReaxFF comparison table.

### Implications for GPU acceleration

Since float32 is not viable, GPU acceleration must focus on:
- **Batch Hessians with `jax.vmap`** (#180) — consolidate N kernel launches into 1
- **End-to-end JIT loss** (#176) — eliminate Python overhead between engine and optimizer
- **JAX-native optimizer** (#176 Phase 3) — replace scipy with JAXopt

Closes #178

### CODE REVIEW

File: q2mm/backends/mm/_jax_common.py [hunk 1]
```diff
@@ -15,11 +16,16 @@
...
+    # JAX defaults to float32.  For MM parameter optimization float64 is the
+    # safe default (energy differences ~1e-6 kcal/mol matter).  However, for
+    # harmonic-only force fields, float32 may be viable and unlocks 64×
+    # throughput on consumer GPUs (see issue #178).
+    #
+    # Honour the standard JAX_ENABLE_X64 env-var: when explicitly set to "0"
+    # we do NOT override, allowing float32 experiments.  Otherwise we enable
+    # float64 as before (standard practice in JAX-based chemistry packages).
+    _user_disabled_x64 = os.environ.get("JAX_ENABLE_X64", "").strip() == "0"
+    if not jax.config.jax_enable_x64 and not _user_disabled_x64:
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/backends/mm/_jax_common.py]: Fixed in 6084047. Now checks `"JAX_ENABLE_X64" in os.environ` — any explicit value (0, false, False, no) is left for JAX to interpret. Only overrides when the env var is completely absent.

File: scripts/float32_experiment.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+    for key in sorted(f64.keys()):
+        if key in ("precision", "throughput_ch3f", "throughput_rh"):
+            continue
+        freqs_64 = np.array(f64[key]["frequencies"])
+        freqs_32 = np.array(f32[key]["frequencies"])
+
+        if len(freqs_64) != len(freqs_32):
+            print(f"  {key}: LENGTH MISMATCH ({len(freqs_64)} vs {len(freqs_32)})")
+            continue
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply scripts/float32_experiment.py]: Fixed in 6084047. Added upfront key validation with warnings for mismatched keys, skip logic for missing systems, and an early return when no frequencies are comparable.

File: docs/benchmarks/gpu.md [hunk 3]
```diff
@@ -289,18 +285,35 @@ later ones.
...
-float32 is not sufficient for final production runs, it may be
-useful for early optimisation cycles where rough convergence is
-acceptable before switching to float64 for refinement.
+### Float32 viability test ✅ Completed
+
+**Result: Float32 is NOT viable for the strict 0.1 cm⁻¹ threshold.**
+Errors up to 0.78 cm⁻¹ appear in real vibrational modes for
+rh-enamide (62 atoms).  Near-zero translation/rotation modes show
+errors up to 9.1 cm⁻¹ due to the eigenvalue decomposition losing
+precision in float32.  CH₃F (5 atoms) passes easily (max 0.0002 cm⁻¹).
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply docs/benchmarks/gpu.md]: This section was completely rewritten in 6084047 with a full methodology section. The new text clearly separates "real modes (>50 cm⁻¹)" from "all modes" in dedicated tables for both full-float32 and mixed-precision experiments.
